dev/test/transaction.ts

```diff
@@ -416,14 +424,51 @@ describe('failed transactions', () => {
     });
   });
 
-  it("doesn't retry on callback failure", () => {
+  it('retries GRPC exceptions with code ABORTED in callback', () => {
+    const retryableError = new GoogleError('Aborted');
+    retryableError.code = Status.ABORTED;
+
+    return runTransaction(
+      async (transaction, docRef) => {
+        await transaction.get(docRef);
+        return 'success';
+      },
+      begin('foo1'),
+      getDocument('foo1', retryableError),
+      rollback('foo1'),
+      begin('foo2', 'foo1'),
+      getDocument('foo2'),
+      commit('foo2')
+    ).then(red => {
+      expect(red).to.equal('success');
+    });
+  });
+
+  it("doesn't retry GRPC exceptions with code FAILED_PRECONDITION in callback", () => {
+    const nonRetryableError = new GoogleError('Failed Precondition');
+    nonRetryableError.code = Status.FAILED_PRECONDITION;
+
+    return expect(
+      runTransaction(
+        async (transaction, docRef) => {
+          await transaction.get(docRef);
+          return 'failure';
+        },
+        begin('foo'),
```

Comment:
If 'foo' is default, do we need it here as well, or can it also be removed?

Reply:
I left it as is here since `getDocument` uses the second argument. I prefer this over only specifying the transaction ID once or using `undefined` in its place.